libgadget/metal_return.c

```diff
@@ -1021,3 +1021,435 @@ stellar_density(const ActiveParticles * act, MyFloat * StellarAges, MyFloat * Ma
+        /* Walk through the new list until we find someone not in the original list.
+         * Add them, then continue.*/
+        int i;
+        for(i = 0; i < remote->nnb; i++) {
+            /* Check whether the list changed: if the furthest particle is still the same we have no merging to do.*/
+            if(remote->nnb == STELLAR_KNN_GET_PRIV(tw)->nnb[pi]
+                    && remote->neighbours[remote->nnb-1].part == neigh[remote->nnb-1].part)
+                break;
+            const struct Neighbour nn = remote->neighbours[i];
+            if(nn.part == neigh[i].part)
+                continue;
+            STELLAR_KNN_GET_PRIV(tw)->nnb[pi] = insert_element(nn.part, nn.dist2, neigh, STELLAR_KNN_GET_PRIV(tw)->nnb[pi]);
+        }
+    }
+}
+
+static void
+stellar_knn_ngbiter(
+        TreeWalkQueryStellarKNN * I,
+        TreeWalkResultStellarKNN * O,
+        TreeWalkNgbIterStellarKNN * iter,
+        LocalTreeWalk * lv)
+{
+    if(iter->base.other == -1) {
+        iter->base.Hsml = lv->tw->tree->BoxSize;
+        O->nnb = I->nnb;
+        if(I->nnb > 0) {
+            memmove(O->neighbours, I->neighbours, sizeof(I->neighbours[0])* I->nnb);
+        }
+        /* If the list is full, we can cull aggressively.
+         * Usually this is secondary treewalks.*/
+        if(I->nnb == NUMNB)
+            iter->base.Hsml = sqrt(I->neighbours[I->nnb-1].dist2);
+        iter->base.mask = 1; /* gas only */
+        iter->base.symmetric = NGB_TREEFIND_ASYMMETRIC;
+        return;
+    }
+    const int other = iter->base.other;
+    const double r2 = iter->base.r2;
+
+    /* Wind particles do not interact hydrodynamically: don't receive metal mass.*/
+    if(winds_is_particle_decoupled(other))
+        return;
+    if(O->nnb == NUMNB && r2 >= O->neighbours[O->nnb-1].dist2)
+        return;
+    O->nnb = insert_element(other, r2, O->neighbours, O->nnb);
+    /* Adjust the treewalk Hsml so we can cull more nodes*/
+    if(O->nnb == NUMNB)
+        iter->base.Hsml = sqrt(O->neighbours[O->nnb-1].dist2);
+}
+
+#define FACT1 0.366025403785	/* FACT1 = 0.5 * (sqrt(3)-1) */
+
+/**
+ * Cull a node.
+ *
+ * Returns 1 if the node shall be opened;
+ * Returns 0 if the node has no business with this query.
+ */
+static int
+cull_node(const TreeWalkQueryBase * const I, const TreeWalkNgbIterBase * const iter, const struct NODE * const current, const double BoxSize)
+{
+    double dist = iter->Hsml + 0.5 * current->len;
+
+    double r2 = 0;
+    double dx = 0;
+    /* do each direction */
+    int d;
+    for(d = 0; d < 3; d ++) {
+        dx = NEAREST(current->center[d] - I->Pos[d], BoxSize);
+        if(dx > dist) return 0;
+        if(dx < -dist) return 0;
+        r2 += dx * dx;
+    }
+    /* now test against the minimal sphere enclosing everything */
+    dist += FACT1 * current->len;
+
+    if(r2 > dist * dist) {
+        return 0;
+    }
+    return 1;
+}
+/*****
+ * This is the internal code that looks for particles in the ngb tree from
```

Comment:
Is this section updated for the knn search?

Reply:
No, let me fix it. I copy-pasted more code than I actually needed.